sycl/doc/extensions/proposed/sycl_ext_oneapi_root_group.asciidoc

```diff
@@ -143,66 +143,139 @@ supports.
+already have equivalents in the `kernel_device_specific` or `device`
+namespaces.
+
+NOTE: Redefining these queries is necessary to define `Param::argument_types`.
+If this extension is proposed for a future SYCL standard, such duplication
+could be avoided.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::info::kernel {
+
+template <uint32_t Dimensions>
+struct max_work_item_sizes;
+
+struct max_work_group_size;
+struct max_num_work_groups;
+
+struct max_sub_group_size;
+struct max_num_sub_groups;
 
-[%header,cols="1,5,5"]
+}
+----
+
+[%header,cols="1,5,5,5"]
 |===
 |Kernel Descriptor
+|Argument Types
 |Return Type
 |Description
 
-|`kernel_queue_specific::max_num_work_group_sync`
+|`template <uint32_t Dimensions>
+ max_work_item_sizes`
+|`sycl::queue`
+|`id<Dimensions>`
+|Returns the maximum number of work-items that are permitted in each dimension
+ of a work-group, when the kernel is submitted to the specified queue,
+ accounting for any kernel properties or features. If the kernel can be
+ submitted to the specified queue without an error, the minimum value returned
+ by this query is 1, otherwise it is 0.
+
+|`max_work_group_size`
+|`sycl::queue`
 |`size_t`
-|Return the maximum number of work-groups that can be synchronized by passing
-a `root_group` to group functions and algorithms. Must be at least 1.
+|Returns the maximum number of work-items that are permitted in a work-group,
+when the kernel is submitted to the specified queue, accounting for any
+kernel properties or features. If the kernel can be submitted to the specified
+queue without an error, the minimum value returned by this query is 1,
+otherwise it is 0.
+
+|`max_num_work_groups`
+|`sycl::queue`, `sycl::range`, `size_t`
+|`size_t`
+|Returns the maximum number of work-groups, when the kernel is submitted to the
+specified queue with the specified work-group size and the specified amount of
+dynamic work-group local memory (in bytes), accounting for any kernel
+properties or features. If the kernel can be submitted to the specified queue
+without an error, the minimum value returned by this query is 1, otherwise it
+is 0.
+
+|`max_sub_group_size`
+|`sycl::queue`, `sycl::range`
+|`uint32_t`
+|Returns the maximum sub-group size, when the kernel is submitted to the
+specified queue with the specified work-group size, accounting for any kernel
+properties or features. If the kernel can be submitted to the specified queue
+without an error, the minimum value returned by this query is 1, otherwise it
+is 0.
+
+|`max_num_sub_groups`
+|`sycl::queue`, `sycl::range`
+|`uint32_t`
+|Returns the maximum number of sub-groups per work-group, when the kernel is
+submitted to the specified queue with the specified work-group size, accounting
+for any kernel properties or features. If the kernel can be submitted to the
+specified queue without an error, the minimum value returned by this query is
+1, otherwise it is 0.
```

Comment:
Why do we want to duplicate these queries in both extensions?

Philosophically, I'm not opposed to duplicating an API in more than one extension.  This just means that the API is available in implementations that support *any* of the extensions.  Of course, the API needs to be defined identically in them all.

Practically, this technique makes more sense for small APIs, though, because there is a larger danger that a large shared API will get out-of-sync between the extensions that define it.  This set of five queries is starting to get on the large size.  Is it important to define them in both extensions?  Would it make more sense to move these queries someplace common like [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/d7dbf7d98ef2eec54aceedb579181530ff8fa2f6/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc)?

Reply:
I duplicated things because I thought it would be easier than adding another dependency.  I don't think the forward progress extension should be dependent on the root-group extension (or vice versa).

sycl_ext_oneapi_kernel_properties is already experimental, and so I thought adding these new interfaces (but not implementing them) would be complicated.  I thought convention was to either add a note that the new interfaces weren't supported, or duplicate the extension into proposed.  Both seemed confusing for this case.

I'm not strongly opposed to moving things into something like a sycl_ext_oneapi_launch_queries extension, though, if you think that would be better.  That would split things out, but still allow the root-group and forward progress extensions to be implemented independently.

Reply:
I have a weak preference for moving these queries to a new common extension, rather than duplicating them in both sycl_ext_oneapi_forward_progress and sycl_ext_oneapi_root_group.

Reply:
Done in ec5f9980fb91e6985aec0762ac4cdf4df97128f4.